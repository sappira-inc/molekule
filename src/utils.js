import { get, kebabCase } from 'lodash';
import styled from 'styled-components';

export const themeGet = (theme, lookup, fallback) => get(theme, lookup, fallback);

export const getComponentVariant = (theme, componentName, variant) => {
  const config = themeGet(
    theme,
    `components.${componentName}.variants.${variant}`,
    theme.variants[componentName][variant]
  );
  if (!config) throw new Error(`Molekule: "${variant}" variant not found in theme...`);
  return config;
};

export const getComponentStyle = componentName => ({ theme }) => themeGet(theme, `components.${componentName}.style`);

export const getComponentClassName = ({ className, theme: { classPrefix }, variant }, name) =>
  `${className || ''} ${classPrefix}-${name} ${variant ? `${classPrefix}-${name}-${variant}` : ''}`.trim();

export const createComponent = ({ name, tag = 'div', as, style, props: defaultProps = () => ({}) }) => {
  const component = as ? styled(as) : styled[tag];

  return component.attrs(props => ({
    ...defaultProps(props),
    ...props,
    className: getComponentClassName(props, kebabCase(name)),
  }))`
    ${style}
    ${getComponentStyle(name)}
    ${({ styles = {} }) => styles[name] || {}}
  `;
};
