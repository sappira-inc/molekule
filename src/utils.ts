import kebabCase from 'lodash/kebabCase';
import get from 'lodash/get';
import styled, { CSSProp } from 'styled-components';

export const themeGet = (lookup: any, fallback?: any) => ({ theme }: any = {}) => get(theme, lookup, fallback);

export const getComponentVariant = (theme: any, componentName: string, variant: string) => {
  const config = themeGet(`variants.${componentName}.${variant}`)({
    theme,
  });

  if (!config) throw new Error(`Molekule: "${variant}" variant not found in theme...`);
  return config;
};

export const getComponentSize = (theme: any, componentName: string, size: string) =>
  themeGet(`sizes.${componentName}.${size}`, {})({ theme });

export const getComponentStyle = (componentName: string) => themeGet(`styles.${componentName}`, {});

export const getVariantStyles = (componentName: string, variant: string) =>
  themeGet(`variants.${componentName}.${variant}.style`, {});

interface GetComponentClassNameProps {
  className: string;
  theme: {
    classPrefix: string;
  };
  variant: string;
}

const getComponentClassName = (
  { className, theme: { classPrefix }, variant }: GetComponentClassNameProps,
  name: string
) => `${className || ''} ${classPrefix}-${name} ${variant ? `${classPrefix}-${name}-${variant}` : ''}`.trim();

interface CreateComponentProps<T> {
  name: string;
  tag?: keyof JSX.IntrinsicElements | React.JSXElementConstructor<any>;
  as?: React.ComponentType<any>;
  style?: ((props: any) => CSSProp) | CSSProp;
  props?: (props: T) => any;
}

export const createComponent = <
  T extends object,
  O extends keyof JSX.IntrinsicElements | React.ComponentType<any> = 'div'
>({
  name,
  tag = 'div',
  as,
  style,
  props: getBaseProps = () => ({}),
}: CreateComponentProps<T>) => {
  const component = styled<O>((as || tag) as any);

  return component.attrs<T>((props: any) => {
    const baseProps = getBaseProps(props);
    const finalProps = {
      ...baseProps,
      ...props,
      className: baseProps.className || props.className,
    };

    return {
      ...finalProps,
      className: getComponentClassName(finalProps, kebabCase(name)),
    };
  })<T>`
  ${style}
  ${getComponentStyle(name)}
  ${(p: any) => getVariantStyles(name, p.variant)}
  ${({ styles = {} }) => styles[name] || {}}
`;
};

// eslint-disable-next-line
const focusableFilter = (node: any) =>
  node.tabIndex >= 0 && !node.hasAttribute('disabled') ? NodeFilter.FILTER_ACCEPT : NodeFilter.FILTER_SKIP;

type FilterFn = (node?: any) => any;

const createTreeWalker = (root: any, currentNode: any, filterFn: FilterFn = () => NodeFilter.FILTER_ACCEPT) => {
  const treeWalker = document.createTreeWalker(
    root,
    NodeFilter.SHOW_ELEMENT,
    {
      acceptNode: filterFn,
    },
    false
  );
  treeWalker.currentNode = currentNode;
  return treeWalker;
};

export const findPreviousFocusableElement = (root: any, currentNode: any) => {
  const treeWalker = createTreeWalker(root, currentNode, focusableFilter);

  if (!treeWalker.previousNode() || treeWalker.currentNode === root) {
    return treeWalker.lastChild();
  }

  return treeWalker.currentNode;
};

export const findNextFocusableElement = (root: any, currentNode: any) => {
  const treeWalker = createTreeWalker(root, currentNode, focusableFilter);

  if (!treeWalker.nextNode() || treeWalker.currentNode === root) {
    treeWalker.currentNode = root;
    return treeWalker.firstChild();
  }

  return treeWalker.currentNode;
};

export const getNextCursorPosition = (cursorPos: any, oldValue: any, newValue: any) => {
  const cursorDiff = newValue.length - oldValue.length;
  let nextPosition = cursorPos;

  if (cursorDiff > 1 || cursorDiff < -1) {
    nextPosition += cursorDiff < 0 ? cursorDiff + 1 : cursorDiff - 1;
  }

  return Math.max(nextPosition, 0);
};
