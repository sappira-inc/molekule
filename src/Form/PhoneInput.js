import React, { useState, useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import { AsYouType, isSupportedCountry, getCountryCallingCode, parseDigits } from 'libphonenumber-js/min';
import examplePhoneNumbers from 'libphonenumber-js/examples.mobile.json';
import Input from './Input';
import { createEasyInput } from './EasyInput';
import { getNextCursorPosition } from '../utils';

export const getRawMaxLength = (countryCode, value) => {
  const countryCallingCode = getCountryCallingCode(countryCode);
  const beginsWithCountryCode = parseDigits(value).substr(0, countryCallingCode.length) === countryCallingCode;
  const examplePhoneNumber = examplePhoneNumbers[countryCode];
  return beginsWithCountryCode ? countryCallingCode.length + examplePhoneNumber.length : examplePhoneNumber.length;
};

function PhoneInput({ countryCode, forwardedRef, value: propValue, onKeyDown, onChange, ...inputProps }) {
  const countryCodeSupported = isSupportedCountry(countryCode);
  if (!countryCodeSupported) {
    throw new Error(`${countryCode} is not supported`);
  }

  const format = (value = '') => {
    const parsed = parseDigits(value).substr(0, getRawMaxLength(countryCode, value));
    return new AsYouType(countryCode).input(parsed);
  };

  const [currentValue, setValue] = useState(format(propValue));
  const inputRef = forwardedRef || useRef();

  useEffect(() => {
    if (propValue !== currentValue) {
      setValue(format(propValue));
    }
  }, [propValue]);

  const handleKeyDown = event => {
    const isLetterLike = /^\w{1}$/.test(event.key);
    if (isLetterLike) {
      const cursorPos = event.target.selectionStart;
      const nextFormattedValue = format(
        currentValue.substring(0, cursorPos) + event.key + currentValue.substring(cursorPos)
      );
      const nextValueRawLength = parseDigits(currentValue).length + 1;
      if (nextValueRawLength > getRawMaxLength(countryCode, nextFormattedValue)) {
        event.preventDefault();
        event.stopPropagation();
      }
    }

    if (onKeyDown) {
      onKeyDown(event);
    }
  };

  const handleChange = (name, newValue, event) => {
    const nextValue = newValue.length < currentValue.length ? newValue.trim() : format(newValue);
    const nextCursorPosition = getNextCursorPosition(event.target.selectionStart, currentValue, nextValue);

    setValue(nextValue);
    setTimeout(() => {
      inputRef.current.setSelectionRange(nextCursorPosition, nextCursorPosition);
    });

    if (onChange) {
      onChange(name, nextValue, event);
    }
  };

  return (
    <Input
      type="tel"
      forwardedRef={inputRef}
      value={currentValue}
      onKeyDown={handleKeyDown}
      onChange={handleChange}
      {...inputProps}
    />
  );
}

PhoneInput.propTypes = {
  ...Input.propTypes,
  countryCode: PropTypes.string,
};

PhoneInput.defaultProps = {
  countryCode: 'US',
};

export default createEasyInput(PhoneInput);
