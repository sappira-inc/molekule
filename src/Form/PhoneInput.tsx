import React, { useState, useEffect, useRef, FC } from 'react';
import { AsYouType, isSupportedCountry, getCountryCallingCode, parseDigits, CountryCode } from 'libphonenumber-js/min';
// @ts-ignore
import examplePhoneNumbers from 'libphonenumber-js/examples.mobile.json';
import { Input, InputProps } from './Input';
import { createEasyInput } from './EasyInput';
import { getNextCursorPosition } from '../utils';

export const getRawMaxLength = (countryCode: any, value: any) => {
  const countryCallingCode = getCountryCallingCode(countryCode);
  const beginsWithCountryCode = parseDigits(value).substr(0, countryCallingCode.length) === countryCallingCode;
  const examplePhoneNumber = examplePhoneNumbers[countryCode];
  return beginsWithCountryCode ? countryCallingCode.length + examplePhoneNumber.length : examplePhoneNumber.length;
};

export interface PhoneInputProps extends InputProps {
  countryCode?: CountryCode;
}

export const PhoneInput: FC<PhoneInputProps> = ({
  countryCode,
  forwardedRef,
  value: propValue,
  onKeyDown,
  onChange,
  ...inputProps
}) => {
  const countryCodeSupported = countryCode && isSupportedCountry(countryCode);
  if (!countryCodeSupported) {
    throw new Error(`${countryCode} is not supported`);
  }

  const format = (value: any) => {
    const phoneString = value || '';
    const parsed = parseDigits(phoneString).substr(0, getRawMaxLength(countryCode, phoneString));
    return new AsYouType(countryCode).input(parsed);
  };

  const [currentValue, setValue] = useState(format(propValue));
  const ref = useRef();
  const inputRef = forwardedRef || ref;

  useEffect(() => {
    if (propValue !== currentValue) {
      setValue(format(propValue));
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [propValue]);

  const handleKeyDown = (event: any) => {
    const isLetterLike = /^\w{1}$/.test(event.key);
    if (isLetterLike) {
      const cursorPos = event.target.selectionStart;
      const nextFormattedValue = format(
        currentValue.substring(0, cursorPos) + event.key + currentValue.substring(cursorPos)
      );
      const nextValueRawLength = parseDigits(currentValue).length + 1;
      if (nextValueRawLength > getRawMaxLength(countryCode, nextFormattedValue)) {
        event.preventDefault();
        event.stopPropagation();
      }
    }

    if (onKeyDown) {
      onKeyDown(event);
    }
  };

  const handleChange = (name: any, newValue: any, event: any) => {
    const nextValue = newValue.length < currentValue.length ? newValue.trim() : format(newValue);
    const nextCursorPosition = getNextCursorPosition(event.target.selectionStart, currentValue, nextValue);

    setValue(nextValue);
    setTimeout(() => {
      inputRef.current.setSelectionRange(nextCursorPosition, nextCursorPosition);
    });

    if (onChange) {
      onChange(name, nextValue, event);
    }
  };

  return (
    <Input
      type="tel"
      forwardedRef={inputRef}
      value={currentValue}
      onKeyDown={handleKeyDown}
      onChange={handleChange}
      {...inputProps}
    />
  );
};

PhoneInput.defaultProps = {
  countryCode: 'US',
};

export default createEasyInput(PhoneInput);
