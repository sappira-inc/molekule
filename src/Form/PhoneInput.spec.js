import React from 'react';
import { renderWithTheme, fireEvent, act } from '../../test/utils';
import PhoneInput from './PhoneInput';

describe('<PhoneInput />', () => {
  const renderInput = props => {
    const utils = renderWithTheme(<PhoneInput placeholder="Input" {...props} />);
    return {
      ...utils,
      input: utils.getByPlaceholderText('Input'),
    };
  };

  const updateInputValue = (input, value) => {
    act(() => {
      fireEvent.change(input, { target: { value } });
    });
  };

  test('snapshot', () => {
    const { asFragment } = renderInput();
    expect(asFragment()).toMatchSnapshot();
  });

  test('initally sets the value', () => {
    const value = '4087213456';
    const { input } = renderInput({ value });

    expect(input.value).toEqual('(408) 721-3456');
  });

  test('area code and hyphens', () => {
    const { input } = renderInput({ value: '408' });

    expect(input.value).toEqual('(408)');

    updateInputValue(input, '(408) 7214');
    expect(input.value).toEqual('(408) 721-4');
  });

  test('can delete paren', () => {
    const { input } = renderInput({ value: '(408)' });

    updateInputValue(input, '(408');
    expect(input.value).toEqual('(408');
  });
});
