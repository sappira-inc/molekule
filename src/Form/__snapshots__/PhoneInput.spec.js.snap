// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PhoneInput /> snapshot 1`] = `
<DocumentFragment>
  .c1 {
  position: relative;
  -webkit-transition: 175ms;
  transition: 175ms;
}

.c1 + .c0 {
  margin-top: 1rem;
}

.c2 {
  position: relative;
}

.c2 + .c4 {
  margin-left: 8px;
}

.c3 {
  border: 1px solid #E8E9EA;
  height: 48px;
  display: block;
  outline: none;
  width: 100%;
  padding: 8px;
  padding-right: 24px;
  border-radius: 8px;
  -webkit-transition: 250ms all;
  transition: 250ms all;
  -webkit-appearance: none;
  font-family: inherit;
  font-size: 16px;
  color: #1B202B;
  box-sizing: border-box;
}

.c3:hover,
.c3:active,
.c3 + label .c3:hover {
  border-color: #A4A6AA;
}

.c3:focus {
  border-color: #0958F3;
}

.c3::-webkit-input-placeholder {
  color: #767980;
  opacity: 1;
}

.c3::-moz-placeholder {
  color: #767980;
  opacity: 1;
}

.c3:-ms-input-placeholder {
  color: #767980;
  opacity: 1;
}

.c3::placeholder {
  color: #767980;
  opacity: 1;
}

.c3[disabled] {
  background-color: #FFFFFF;
  border-color: #E8E9EA;
  color: #D1D2D5;
}

.c3[disabled]::-webkit-input-placeholder {
  color: #D1D2D5;
}

.c3[disabled]::-moz-placeholder {
  color: #D1D2D5;
}

.c3[disabled]:-ms-input-placeholder {
  color: #D1D2D5;
}

.c3[disabled]::placeholder {
  color: #D1D2D5;
}

<div
    class="c0 c1 re-field"
  >
    <div
      class="c2 re-input-container"
    >
      <input
        class="c3 re-input"
        placeholder="Input"
        rows="3"
        type="tel"
        value=""
      />
    </div>
  </div>
</DocumentFragment>
`;
